### PR TITLE
Make the Reductionops conversion functions evarmap-aware.

### PR DESCRIPTION
**Pull request by ppedrot**
This implicitly gives the stability by unification state in #16903 in a completely orthogonal way. Additionally it also untangles some spaghetti code.

### CODE REVIEW

File: pretyping/reductionops.ml [hunk 1]
```diff
@@ -1040,43 +1040,83 @@ let report_anomaly (e, info) =
...
-    let env = Environ.set_universes (Evd.universes sigma) env in
-    try f ~reds:ts env ~evars:(Evd.evar_handler sigma) x y; true
-    with Reduction.NotConvertible -> false
-    | UGraph.UniverseInconsistency _ -> false
-    | e ->
-      let e = Exninfo.capture e in
-      report_anomaly e
+  try generic_conv pb ~reds:ts env sigma x y; true
+  with Reduction.NotConvertible -> false
+  | UGraph.UniverseInconsistency _ -> false
```
Comment (ppedrot) [inline pretyping/reductionops.ml:1116]: cc @SkySkimmer : after carefully looking at the code I think that this try-with branch is actually dead code.  The `generic_conv` function doesn't seem to raise this exception. If so I can even polish this PR a bit more by factorizing this function with the one above.

Comment (SkySkimmer) [reply pretyping/reductionops.ml:1116]: I think that's correct, kernel conversion turns univ inconsstency into notconvertible

Comment (ppedrot) [reply pretyping/reductionops.ml:1116]: OK, I'll force-push the factorization then.

**Comment by ppedrot** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**